backend/test_observer/controllers/test_executions/end_test.py

```diff
@@ -0,0 +1,181 @@
+    artefact = test_execution.artefact_build.artefact
+    prev_artefact = _get_previous_artefact(db, artefact)
+
+    if prev_artefact is None:
+        return None
+
+    query = (
+        select(TestExecution)
+        .join(ArtefactBuild)
+        .where(
+            ArtefactBuild.artefact_id == prev_artefact.id,
+            TestExecution.environment_id == test_execution.environment_id,
+        )
+        .order_by(ArtefactBuild.revision.desc())
+        .limit(1)
+        .options(
+            joinedload(TestExecution.test_results).joinedload(TestResult.test_case)
+        )
+    )
+
+    return db.execute(query).unique().scalar_one_or_none()
+
+
+def _get_previous_artefact(db: Session, artefact: Artefact) -> Artefact | None:
+    family = artefact.stage.family
+
+    query = select(Artefact).where(
+        Artefact.name == artefact.name, Artefact.id != artefact.id
+    )
+
+    if family.name == FamilyName.SNAP.value:
+        query = query.where(Artefact.track == artefact.track)
+
+    if family.name == FamilyName.DEB.value:
+        query = query.where(
+            Artefact.series == artefact.series, Artefact.repo == artefact.repo
+        )
+
+    query = query.order_by(Artefact.id.desc()).limit(1)
+
+    return db.execute(query).scalar_one_or_none()
+
+
+def _find_related_test_execution(
+    request: EndTestExecutionRequest, db: Session
+) -> TestExecution | None:
+    return (
+        db.execute(
+            select(TestExecution)
+            .where(TestExecution.ci_link == request.ci_link)
+            .options(
+                joinedload(TestExecution.artefact_build)
+                .joinedload(ArtefactBuild.artefact)
+                .joinedload(Artefact.stage)
+                .joinedload(Stage.family)
+            )
+            .options(
+                joinedload(TestExecution.test_results).joinedload(TestResult.test_case)
+            )
+        )
+        .unique()
+        .scalar_one_or_none()
+    )
+
+
+def _store_test_results(
+    db: Session,
+    c3_test_results: list[C3TestResult],
+    test_execution: TestExecution,
+) -> None:
+    for r in c3_test_results:
+        test_case = get_or_create(
+            db,
+            TestCase,
+            filter_kwargs={"name": r.name},
+            creation_kwargs={"category": r.category},
+        )
+
+        test_result = TestResult(
+            test_case=test_case,
+            test_execution=test_execution,
+            status=_parse_c3_test_result_status(r.status),
+            comment=r.comment,
+            io_log=r.io_log,
+        )
+
+        db.add(test_result)
+
+    db.commit()
+
+
+def _parse_c3_test_result_status(status: C3TestResultStatus) -> TestResultStatus:
+    match status:
+        case C3TestResultStatus.PASS:
+            return TestResultStatus.PASSED
+        case C3TestResultStatus.FAIL:
+            return TestResultStatus.FAILED
+        case C3TestResultStatus.SKIP:
+            return TestResultStatus.SKIPPED
```

Comment:
Not sure which approach is more effective, but you can just construct a dictionary here like `{c3_result: to_result}` and return `mapped_results[status]`

Reply:
the match/case approach works better with mypy. Since the return type is not optional, mypy would complain if we missed a case. Using a dictionary on the other hand would not give us the same guarantees.